include/fmt/format.h

```diff
@@ -776,7 +790,34 @@ class truncating_iterator {
     return it;
   }
 
-  reference operator*() const { return count_ < limit_ ? *out_ : blackhole_; }
+  reference operator*() const {
+    return this->count_ < this->limit_ ? *this->out_ : blackhole_;
+  }
+};
+
+template <typename OutputIt>
+class truncating_iterator<OutputIt, std::true_type>:
+  public truncating_iterator_base<OutputIt> {
+ public:
+  typedef typename OutputIt::container_type::value_type value_type;
+  typedef void difference_type;
+  typedef void pointer;
+  typedef void reference;
```

Comment:
Why are these `void` and are they different from their counterparts in `iterator_traits`?

Reply:
`void` actually **is** their type in [`iterator_traits`](https://en.cppreference.com/w/cpp/iterator/back_insert_iterator). We are talking about a `back_insert_operator` here. I've basically just redefined the `value_type` from `void` to the `value_type`of the underlying container.

Reply:
Then they can be moved to `truncating_iterator_base` I think.

Reply:
I agree, even if it loses information about the underlying iterator type. Anyway, I've refactored these three `void` typedefs into the common base class, the rest must stay as is: these are the essential differences between "regular output iterators" and "back insert interators".